pallets/parachain-system/src/lib.rs

```diff
@@ -428,6 +547,7 @@ pub mod pallet {
 			<RelayStateProof<T>>::put(relay_chain_state);
 			<RelevantMessagingState<T>>::put(relevant_messaging_state.clone());
 			<HostConfiguration<T>>::put(host_config);
+			<PendingDownwardMessages<T>>::put(downward_messages.clone());
```

Comment:
If it's cleared in `on_finalize` then it never hits storage in practice, only the in-memory overlay.

Reply:
It's inserted in `set_validation_data`, fetched in `on_finalize`, cleared in `on_initialize`

Reply:
two things stand out to me:
  * only the length of the list is used, not the messages themselves
  * it could be cleared in `on_finalize` too, right?

Reply:
it also doesn't seem well-defined to me what `pending_downward_messages` _is_ in this context.

i.e. is it all pending downward messages that were present in the relay-parent state? If so, then we need to update `process_inbound_downward_messages` to avoid processing them in duplicate.

or is it only the pending downward messages which haven't been processed? In that case, we shouldn't have to make any changes to the processing logic.

For efficiency's sake, I'd favor the latter - we shouldn't include the same downward messages into parachain blocks multiple times.

horizontal logic probably needs the same, no?

Reply:
I followed the pattern of cleaning everything in `on_initialize`

As for the messages, yes, I suggested in the first comments to either put only len or 
> Second option is to introduce migration and save its len along with ProcessedDownwardMessages and rename it into something like DownwardQueueSize.

Reply:
> i.e. is it all pending downward messages that were present in the relay-parent state? If so, then we need to update process_inbound_downward_messages to avoid processing them in duplicate.

These are all messages from the downward queue. They're not used for processing in the current code, so it's not an issue. We should probably just use len.
But the thing is, according to the code each candidate exhaust the dm queue:
```rust
let dm_count = downward_messages.len() as u32;
// ...
ProcessedDownwardMessages::<T>::put(dm_count);
```
Do we even need to track this limitation at all?

Reply:
If we assume that `ParachainInherentData` only contains messages which have _not_ been posted to any prior parachain block (which is the way it should be to avoid redundancy), I don't know that that current bandwidth limits approach works.

The bandwidth limits for DMP/HRMP-in should be based on the _total_ size of the incoming queues at the relay-parent, whereas some (or most) of the messages may have already been processed.

Reply:
> according to the code each candidate exhaust the dm queue

Yes, this looks currently guaranteed. We could avoid tracking this property as a result, and reintroduce the concept if we get to the point where we implement things more piecemeal.